### PR TITLE
[MAT-1783] GHA to check if there are untracked GQL types

### PR DESCRIPTION
**Pull request by dulishkovych**
[MAT-1783]

### Description

Add a check that will catch untracked changes on GQL types. It means that the branch has some types that were not generated. 

### How to test

- Please check this [PR](https://github.com/toptal/client-portal/pull/5161) for testing.

### Development checks

- [x] Add changeset according to [guidelines](https://github.com/toptal/davinci-github-actions/blob/master/_docs/contribution/changeset-guidelines.md) (if needed)

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-anvil ping reviewers` - Ping FX team for review

</details>

[MAT-1783]: https://toptal-core.atlassian.net/browse/MAT-1783?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .github/actions/action.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+inputs:
+  path-pattern:
+    description: 'git `pathspec` pattern to match GQL types files'
+    required: false
+    default: '*.gql.ts'
+  generate-types-command:
+    description: 'Command to generate GQL types'
+    required: false
+    default: yarn generate:types
```
Comment (dmaklygin) [inline .github/actions/action.yml:12]: According to RFC-29, the command for generating types from the existing schema should be `yarn types:generate`.
Wdyt, shouldn't we change it?

Comment (dulishkovych) [reply .github/actions/action.yml:12]: In cases when that is not true. I know that is standard, but still, on CP, we have different commands. I will change it with my PR, but still I wanted to make it configurable

File: no-untracked-generated-types/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+The list of arguments, that are used in GH Action:
+
+| name                     | type   | required | default             | description                                     |
+| ------------------------ | ------ | -------- | ------------------- | ----------------------------------------------- |
+| `path-pattern`           | string |          | \*.gql.ts           | git `pathspec` pattern to match GQL types files |
+| `generate-types-command` | string |          | yarn generate:types | Command to generate GQL types                   |
+
+### Outputs
+
+Not specified
```
Comment (falsefalse) [inline no-untracked-generated-types/README.md:20]: I think we want to say that `exit 1` fails the check or something

File: no-untracked-generated-types/action.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+runs:
+  using: composite
+  steps:
+    - shell: bash
+      run: ${{ inputs.generate-types-command }}
+
+    - name: Check for untracked files
+      shell: bash
+      run: |
+        UNTRACKED_FILES_COUNT=0
```
Comment (falsefalse) [inline no-untracked-generated-types/action.yml:23]: But wait, we are interested in files that are changed, not only untracked.

File: no-untracked-generated-types/action.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        git status -s -u ${{ inputs.davinci-branch }} ':(exclude)untracked_files.txt' > untracked_files.txt
+        UNTRACKED_FILES_COUNT=`wc -l untracked_files.txt | cut -c1`
+        if [ $UNTRACKED_FILES_COUNT -gt 0 ]; then
+          echo "Untracked files found:"
+          cat untracked_files.txt
```
Comment (falsefalse) [inline no-untracked-generated-types/action.yml:28]: Would be really nice if it also said what to do about it, as in "Generated types are not in sync, please run yarn types:generate and commit the result" or something along those lines.

File: no-untracked-generated-types/action.yml (hunk 3 above)
Comment (dulishkovych) [reply no-untracked-generated-types/action.yml:23]: On CI, when we do regenerate types, all files that were regenerated would be untracked or am I mistaken? :thinking:

Comment (falsefalse) [reply no-untracked-generated-types/action.yml:23]: You are :) "Add/remove field to existing query/frag/mutation — forget to regen types"

**Review by LashaJini — APPROVED** [review]
Code looks good. Didn't test it. Could you also add changeset, please :bow:

Thank you for contribution :+1:

**Review by TomasSlama — APPROVED** [review]
Code looks good 👍 didn't test

File: no-untracked-generated-types/README.md [hunk 5]
```diff
@@ -0,0 +1,30 @@
+## Check untracked GQL types
+
+Check if some GQL types are not generated
```
Comment (falsefalse) [inline no-untracked-generated-types/README.md:3]:
```suggestion
Check if some GQL types were not generated
```

File: no-untracked-generated-types/README.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+### Description
+
+Checks if there are some queries or mutation which types are outdated/not generated. Runs command `yarn generate:types` and checks if there are some untracked changes.
```
Comment (falsefalse) [inline no-untracked-generated-types/README.md:7]: 👍

File: no-untracked-generated-types/action.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
+name: Check untracked GQL types
+description: Check if some GQL types are not generated
```
Comment (falsefalse) [inline no-untracked-generated-types/action.yml:2]:
```suggestion
description: Check if some GQL types were not generated
```

File: no-untracked-generated-types/action.yml [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    - name: Check for untracked files
+      shell: bash
+      run: |
+        UNTRACKED_CHANGES_COUNT=0
+        git status -s -u ${{ inputs.davinci-branch }} ':(exclude)untracked_changes.txt' > untracked_changes.txt
+        UNTRACKED_CHANGES_COUNT=`wc -l untracked_changes.txt | cut -c1`
+        if [ $UNTRACKED_CHANGES_COUNT -gt 0 ]; then
+          echo "Files with untracked changes found:"
+          cat untracked_changes.txt
+          echo "::error::Generated types are not in sync, please run ${{ inputs.generate-types-command }} and commit the result"
```
Comment (falsefalse) [inline no-untracked-generated-types/action.yml:29]: Makes perfect sense, but lacks emojis :)

**Review by falsefalse — APPROVED** [review]
👏

File: no-untracked-generated-types/action.yml (hunk 8 above)
Comment (dulishkovych) [reply no-untracked-generated-types/action.yml:29]: :smile:

**Review by separatio — APPROVED** [review]
Looks good. Didn't test, but checked the PR that runs the query, and saw the failure. Looks good. :+1:

**Pull request merged**